ESBuild SCSS style budget assessments change drastically if source maps are enabled

### Command

build

### Is this a regression?

- [ ] Yes, this behavior used to work in the previous version

### The previous version in which this bug was not present was

_No response_

### Description

When using either of the new esbuild-based builders (`application` and `browser-esbuild`), if SCSS stylesheets and source maps are both used, the calculation of the style file sizes for budget verification results in drastically increased reported sizes.

### Minimal Reproduction

1. Create a new Angular app with `ng new`, using SCSS as the stylesheet format
2. Add the contents of [this file](https://github.com/HiReception/angular-esbuild-bug-demo/blob/esbuild-without-imports/src/app/app.component.scss) to `app.component.scss`
3. Run `ng build` - the build fails because the `app.component.scss` style file exceeds budget; this is expected, but the key note is that the listed size of the style file is 20.17 kB
4. In the production build configuration in `angular.json`, add `"sourceMap": true,`
5. Run `ng build` again - the build again fails due to budget breaches, but the listed size of the style file in question is now 77.26 kB

### Exception or Error

```text
Expected error: 
X [ERROR] src/app/app.component.scss exceeded maximum budget. Budget 4.00 kB was not met by 16.17 kB with a total of 20.17 kB.

Actual error:
X [ERROR] src/app/app.component.scss exceeded maximum budget. Budget 4.00 kB was not met by 73.26 kB with a total of 77.26 kB.
```


### Your Environment

```text
_                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/


Angular CLI: 17.0.6
Node: 20.10.0
Package Manager: npm 8.19.2
OS: win32 x64

Angular: 17.0.6
... animations, cli, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1700.6
@angular-devkit/build-angular   17.0.6
@angular-devkit/core            17.0.6
@angular-devkit/schematics      17.0.6
@schematics/angular             17.0.6
rxjs                            7.8.1
typescript                      5.2.2
zone.js                         0.14.2
```


### Anything else relevant?

_No response_

--- a/packages/angular_devkit/build_angular/src/builders/application/tests/options/sourcemap_spec.ts
+++ b/packages/angular_devkit/build_angular/src/builders/application/tests/options/sourcemap_spec.ts
@@ -136,5 +136,41 @@ describeBuilder(buildApplication, APPLICATION_BUILDER_INFO, (harness) => {
 
       harness.expectFile('dist/browser/main.js.map').content.toContain('"x_google_ignoreList"');
     });
+
+    it('should generate component sourcemaps when sourcemaps when true', async () => {
+      await harness.writeFile('src/app/app.component.css', `* { color: red}`);
+
+      harness.useTarget('build', {
+        ...BASE_OPTIONS,
+        sourceMap: true,
+      });
+
+      const { result } = await harness.executeOnce();
+
+      expect(result?.success).toBeTrue();
+
+      harness
+        .expectFile('dist/browser/main.js')
+        .content.toContain('sourceMappingURL=app.component.css.map');
+      harness.expectFile('dist/browser/app.component.css.map').toExist();
+    });
+
+    it('should not generate component sourcemaps when sourcemaps when false', async () => {
+      await harness.writeFile('src/app/app.component.css', `* { color: red}`);
+
+      harness.useTarget('build', {
+        ...BASE_OPTIONS,
+        sourceMap: false,
+      });
+
+      const { result } = await harness.executeOnce();
+
+      expect(result?.success).toBeTrue();
+
+      harness
+        .expectFile('dist/browser/main.js')
+        .content.not.toContain('sourceMappingURL=app.component.css.map');
+      harness.expectFile('dist/browser/app.component.css.map').toNotExist();
+    });
   });
 });

--- a/packages/angular_devkit/build_angular/src/tools/esbuild/angular/compiler-plugin.ts
+++ b/packages/angular_devkit/build_angular/src/tools/esbuild/angular/compiler-plugin.ts
@@ -164,14 +164,15 @@ export function createCompilerPlugin(
               );
             }
 
-            const { contents, resourceFiles, referencedFiles, errors, warnings } = stylesheetResult;
+            const { contents, outputFiles, metafile, referencedFiles, errors, warnings } =
+              stylesheetResult;
             if (errors) {
               (result.errors ??= []).push(...errors);
             }
             (result.warnings ??= []).push(...warnings);
             additionalResults.set(stylesheetFile ?? containingFile, {
-              outputFiles: resourceFiles,
-              metafile: stylesheetResult.metafile,
+              outputFiles,
+              metafile,
             });
 
             if (referencedFiles) {

--- a/packages/angular_devkit/build_angular/src/tools/esbuild/angular/component-stylesheets.ts
+++ b/packages/angular_devkit/build_angular/src/tools/esbuild/angular/component-stylesheets.ts
@@ -57,7 +57,7 @@ export class ComponentStylesheetBundler {
       });
     });
 
-    return extractResult(await bundlerContext.bundle(), bundlerContext.watchFiles);
+    return this.extractResult(await bundlerContext.bundle(), bundlerContext.watchFiles);
   }
 
   async bundleInline(data: string, filename: string, language: string) {
@@ -103,7 +103,7 @@ export class ComponentStylesheetBundler {
     });
 
     // Extract the result of the bundling from the output files
-    return extractResult(await bundlerContext.bundle(), bundlerContext.watchFiles);
+    return this.extractResult(await bundlerContext.bundle(), bundlerContext.watchFiles);
   }
 
   invalidate(files: Iterable<string>) {
@@ -128,52 +128,50 @@ export class ComponentStylesheetBundler {
 
     await Promise.allSettled(contexts.map((context) => context.dispose()));
   }
-}
 
-function extractResult(result: BundleContextResult, referencedFiles?: Set<string>) {
-  let contents = '';
-  let map;
-  let outputPath;
-  const resourceFiles: OutputFile[] = [];
-  if (!result.errors) {
-    for (const outputFile of result.outputFiles) {
-      const filename = path.basename(outputFile.path);
-      if (outputFile.type === BuildOutputFileType.Media) {
-        // The output files could also contain resources (images/fonts/etc.) that were referenced
-        resourceFiles.push(outputFile);
-      } else if (filename.endsWith('.css')) {
-        outputPath = outputFile.path;
-        contents = outputFile.text;
-      } else if (filename.endsWith('.css.map')) {
-        map = outputFile.text;
-      } else {
-        throw new Error(
-          `Unexpected non CSS/Media file "${filename}" outputted during component stylesheet processing.`,
-        );
+  private extractResult(result: BundleContextResult, referencedFiles?: Set<string>) {
+    let contents = '';
+    let metafile;
+    const outputFiles: OutputFile[] = [];
+
+    if (!result.errors) {
+      for (const outputFile of result.outputFiles) {
+        const filename = path.basename(outputFile.path);
+
+        // Needed for Bazel as otherwise the files will not be written in the correct place.
+        outputFile.path = path.join(this.options.workspaceRoot, outputFile.path);
+
+        if (outputFile.type === BuildOutputFileType.Media) {
+          // The output files could also contain resources (images/fonts/etc.) that were referenced
+          outputFiles.push(outputFile);
+        } else if (filename.endsWith('.css')) {
+          contents = outputFile.text;
+        } else if (filename.endsWith('.css.map')) {
+          outputFiles.push(outputFile);
+        } else {
+          throw new Error(
+            `Unexpected non CSS/Media file "${filename}" outputted during component stylesheet processing.`,
+          );
+        }
       }
+
+      metafile = result.metafile;
+      // Remove entryPoint fields from outputs to prevent the internal component styles from being
+      // treated as initial files. Also mark the entry as a component resource for stat reporting.
+      Object.values(metafile.outputs).forEach((output) => {
+        delete output.entryPoint;
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        (output as any)['ng-component'] = true;
+      });
     }
-  }
 
-  let metafile;
-  if (!result.errors) {
-    metafile = result.metafile;
-    // Remove entryPoint fields from outputs to prevent the internal component styles from being
-    // treated as initial files. Also mark the entry as a component resource for stat reporting.
-    Object.values(metafile.outputs).forEach((output) => {
-      delete output.entryPoint;
-      // eslint-disable-next-line @typescript-eslint/no-explicit-any
-      (output as any)['ng-component'] = true;
-    });
+    return {
+      errors: result.errors,
+      warnings: result.warnings,
+      contents,
+      outputFiles,
+      metafile,
+      referencedFiles,
+    };
   }
-
-  return {
-    errors: result.errors,
-    warnings: result.warnings,
-    contents,
-    map,
-    path: outputPath,
-    resourceFiles,
-    metafile,
-    referencedFiles,
-  };
 }

--- a/packages/angular_devkit/build_angular/src/tools/esbuild/angular/jit-plugin-callbacks.ts
+++ b/packages/angular_devkit/build_angular/src/tools/esbuild/angular/jit-plugin-callbacks.ts
@@ -121,10 +121,10 @@ export function setupJitPluginCallbacks(
         );
       }
 
-      const { contents, resourceFiles, errors, warnings, metafile, referencedFiles } =
+      const { contents, outputFiles, errors, warnings, metafile, referencedFiles } =
         stylesheetResult;
 
-      additionalResultFiles.set(entry.path, { outputFiles: resourceFiles, metafile });
+      additionalResultFiles.set(entry.path, { outputFiles, metafile });
 
       return {
         errors,

--- a/packages/angular_devkit/build_angular/src/tools/esbuild/compiler-plugin-options.ts
+++ b/packages/angular_devkit/build_angular/src/tools/esbuild/compiler-plugin-options.ts
@@ -60,7 +60,7 @@ export function createCompilerPluginOptions(
         // Hidden component stylesheet sourcemaps are inaccessible which is effectively
         // the same as being disabled. Disabling has the advantage of avoiding the overhead
         // of sourcemap processing.
-        !!sourcemapOptions.styles && (sourcemapOptions.hidden ? false : 'inline'),
+        sourcemapOptions.styles && !sourcemapOptions.hidden ? 'linked' : false,
       outputNames,
       includePaths: stylePreprocessorOptions?.includePaths,
       externalDependencies,

--- a/packages/angular_devkit/build_angular/src/tools/esbuild/stylesheets/bundle-options.ts
+++ b/packages/angular_devkit/build_angular/src/tools/esbuild/stylesheets/bundle-options.ts
@@ -22,7 +22,7 @@ export interface BundleStylesheetOptions {
   optimization: boolean;
   inlineFonts: boolean;
   preserveSymlinks?: boolean;
-  sourcemap: boolean | 'external' | 'inline';
+  sourcemap: boolean | 'external' | 'inline' | 'linked';
   outputNames: { bundles: string; media: string };
   includePaths?: string[];
   externalDependencies?: string[];

